src/resolve/resolved-bus.c

```diff
@@ -59,54 +59,69 @@ static int dns_query_bus_track(DnsQuery *q, sd_bus_message *m) {
         return 0;
 }
 
+static sd_bus_message *dns_query_steal_request(DnsQuery *q) {
+        assert(q);
+
+        /* Find the main query, it's the one that owns the message */
+        while (q->auxiliary_for)
+                q = q->auxiliary_for;
+
+        /* Let's take the request message out of the DnsQuery object, so that we never send requests twice */
+        return TAKE_PTR(q->bus_request);
+}
+
 static int reply_query_state(DnsQuery *q) {
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *req = NULL;
 
         assert(q);
-        assert(q->bus_request);
+
+        req = dns_query_steal_request(q);
+        if (!req) /* No bus message set anymore? then we already replied already, let's not answer a second time */
+                return 0;
 
         switch (q->state) {
 
         case DNS_TRANSACTION_NO_SERVERS:
-                return sd_bus_reply_method_errorf(q->bus_request, BUS_ERROR_NO_NAME_SERVERS, "No appropriate name servers or networks for name found");
+                return sd_bus_reply_method_errorf(req, BUS_ERROR_NO_NAME_SERVERS, "No appropriate name servers or networks for name found");
 
         case DNS_TRANSACTION_TIMEOUT:
-                return sd_bus_reply_method_errorf(q->bus_request, SD_BUS_ERROR_TIMEOUT, "Query timed out");
+                return sd_bus_reply_method_errorf(req, SD_BUS_ERROR_TIMEOUT, "Query timed out");
```

Comment:
Shouldn't this instead be:
```C
switch (q->state) {


case DNS_TRANSACTION_NO_SERVERS:
      return reply_method_errorf(q, SD_BUS_ERROR_TIMEOUT, "All attempts to contact name servers or networks failed");
case DNS_TRANSACTION_INVALID_REPLY:
      return reply_method_errorf(q, BUS_ERROR_INVALID_REPLY, "Received invalid reply");
...
```

It seems inconsistent to open-code the call to dns_query_steal_request() here.